token-metadata/interface/src/instruction.rs

```diff
@@ -0,0 +1,413 @@
+//! Instruction types
+
+use {
+    crate::state::OptionalNonZeroPubkey,
+    borsh::{BorshDeserialize, BorshSerialize},
+    solana_program::{
+        instruction::{AccountMeta, Instruction},
+        program_error::ProgramError,
+        pubkey::Pubkey,
+    },
+    spl_type_length_value::discriminator::{Discriminator, TlvDiscriminator},
+};
+
+/// Fields in the metadata account
+#[derive(Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize)]
+pub enum Field {
+    /// The name field, corresponding to `TokenMetadata.name`
+    Name,
+    /// The symbol field, corresponding to `TokenMetadata.symbol`
+    Symbol,
+    /// The uri field, corresponding to `TokenMetadata.uri`
+    Uri,
+    /// A user field, whose key is given by the associated string
+    Key(String),
+}
+
+/// Initialization instruction data
+#[derive(Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize)]
+pub struct Initialize {
+    /// Longer name of the token
+    pub name: String,
+    /// Shortened symbol of the token
+    pub symbol: String,
+    /// URI pointing to more metadata (image, video, etc.)
+    pub uri: String,
+}
+impl TlvDiscriminator for Initialize {
+    /// Please use this discriminator in your program when matching
+    const TLV_DISCRIMINATOR: Discriminator = Discriminator::new(INITIALIZE_DISCRIMINATOR);
+}
+/// First 8 bytes of `hash::hashv(&["spl-token-metadata-interface:metadata-initialize"])`
+const INITIALIZE_DISCRIMINATOR: [u8; Discriminator::LENGTH] =
+    [229, 250, 62, 140, 225, 241, 85, 223];
+// annoying, but needed to perform a match on the value
+const INITIALIZE_DISCRIMINATOR_SLICE: &[u8] = &INITIALIZE_DISCRIMINATOR;
+
+/// Update field instruction data
+#[derive(Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize)]
+pub struct UpdateField {
+    /// Field to update in the metadata
+    pub field: Field,
+    /// Value to write for the field
+    pub value: String,
+}
+impl TlvDiscriminator for UpdateField {
+    /// Please use this discriminator in your program when matching
+    const TLV_DISCRIMINATOR: Discriminator = Discriminator::new(UPDATE_FIELD_DISCRIMINATOR);
+}
+/// First 8 bytes of `hash::hashv(&["spl-token-metadata-interface:update-field"])`
```

Comment:
For the sake of macro-izing things with the toolkit I'm working on, I'd encourage use of underscore `_` instead of hyphen `-` ie. `spl_token_metadata_interface:update_field`.

TL/DR: Defining a macro attribute like `#[interface(spl-token-metadata-interface:update-field)]` causes `syn` to fumble the bag

Reply:
Yeah sure, no problem. I'll probably need to change up the discriminator strings again though to ensure those high first bytes

Reply:
Sounds good, I approved but this change would help out what I'm working on quite a bit!